### PR TITLE
Amqp Filters

### PR DESCRIPTION
**Pull request by patrickdet**
This PR adds the Amqp Filter functionality #3 to the client. Upon attaching a receiver link one can pass a map of filters.

So far I have only written a test for Azure but all the other tests are green so no existing functionality is negatively affected.

Let me know what you think. I need this functionality for an upcoming project which connects to Azure Eventhubs and since people have asked for the ability to set filters I thought it would be a good time to implement them.

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@patrickdet Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/rabbitmq?repositoryId=rabbitmq/rabbitmq-amqp1.0-client&pullRequestId=17)!

[Click here](https://cla.pivotal.io/sync/rabbitmq?repositoryId=rabbitmq/rabbitmq-amqp1.0-client&pullRequestId=17) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@patrickdet Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/rabbitmq?repositoryId=rabbitmq/rabbitmq-amqp1.0-client&pullRequestId=17)!

**Comment by michaelklishin** [comment]
@patrickdet thank you. The lead maintainer of this client will be away for over a week but we have made a note.

File: src/amqp10_client_session.erl
Comment (kjnilsson) [inline src/amqp10_client_session.erl]: This will never match as the above clause does not match on the empty map - it matches on it _being_ a map and thus always passes. You need to use the `map_size` guard function to check for the empty map.

File: src/amqp10_client_session.erl
Comment (kjnilsson) [inline src/amqp10_client_session.erl]: nitpick = unnecessary newline

File: src/amqp10_client_session.erl
Comment (kjnilsson) [inline src/amqp10_client_session.erl]: As we cannot run the azure tests as part of ci I would like some local unit tests for this function

File: test/system_SUITE.erl
Comment (kjnilsson) [inline test/system_SUITE.erl:367]: There are dialyzer errors in this function

```  Checking whether the PLT /Users/knilsson/code/rabbitmq/rabbitmq-public-umbrella-master/deps/amqp10_client/.amqp10_client.plt is up-to-date... yes
  Proceeding with analysis...
system_SUITE.erl:306: Function filtered_roundtrip_service_bus/1 has no local return
system_SUITE.erl:364: Function filtered_roundtrip/1 has no local return
system_SUITE.erl:367: Function filtered_roundtrip/2 has no local return
system_SUITE.erl:400: Call to missing or unexported function amqp10_client:attach_receiver_link/6
system_SUITE.erl:405: Binary construction will fail since the value field Now2Binary in segment Now2Binary:8/integer-unit:1 has type binary()```

Comment (kjnilsson) [reply test/system_SUITE.erl:367]: Looks like you haven't exported `amqp10_client:attach_receiver_link/6`

**Review by kjnilsson — CHANGES_REQUESTED** [review]
Hi,

Thanks for this, it would be a useful feature for the client. There are a few things to address first though. Cheers.

**Comment by patrickdet** [comment]
Thank you @kjnilsson for the pointers. It's been a while since I have gotten to write some Erlang. I have addressed them now. Let me know what you think. 

Btw would it be possible to release the client to hex.pm when/if this feature gets merged?

PS (unrelated): Big fan of [Fez](https://github.com/kjnilsson/fez) 👍

**Comment by michaelklishin** [comment]
@patrickdet a new release can take a couple of days after merging but shouldn't take long. Thank you!

**Review by michaelklishin — APPROVED** [review]

**Review by kjnilsson — APPROVED** [review]
Looks good now.

**Pull request merged**
